### ISSUE TITLE
docker-compose.yml environment variables no longer working

### ISSUE THREAD

**Issue by davidoshiro** [docker/compose#4740]
I'm not sure what is going on but this is pretty frustrating. I had working docker-compose.yml file that I use as a template for creating new projects. Today, I created a new project using the same docker-compose.yml file as my working projects, but when doing a 'docker-compose build' command, I get an error saying that my environment variables do not exist. Of course, they do exists but are not being recognized for some reason. I go through the online docs and it says to use $$ where I reference the environment variables. The build works, but now when I try 'docker-compose up', docker sees the environment value as ${VAR_NAME} instead of the value.

I have Docker Community Edition version 17.05.0-ce-rc1-mac8 (16582).

Any help is appreciated.

**Comment by davidoshiro** [comment]
Could it be that 'docker-compose build' and 'docker-compose up' interprets environment variables in the docker-compose.yml differently? That is what it seems to be happening for me.

When using ${} in docker-compose.yml I get the following when running 'docker-compose build':
WARNING: compose.config.environment.__getitem__: The MYSQL_DB_PORT variable is not set. Defaulting to a blank string.
WARNING: compose.config.environment.__getitem__: The APP_NAME variable is not set. Defaulting to a blank string.
WARNING: compose.config.environment.__getitem__: The archeryscorekeeper variable is not set. Defaulting to a blank string.
WARNING: compose.config.environment.__getitem__: The APP_PORT variable is not set. Defaulting to a blank string.
ERROR: compose.cli.main.main: The Compose file './docker-compose.yml' is invalid because:
services.db.ports is invalid: Invalid port ":", should be [[remote_ip:]remote_port[-remote_port]:]port[/protocol]
services.rails.ports is invalid: Invalid port ":", should be [[remote_ip:]remote_port[-remote_port]:]port[/protocol]

When using $${} the build succeeds. But, when I try to run 'docker-compose run' I get the following:
Could it be that 'docker-compose build' and 'docker-compose up' interprets environment variables in the docker-compose.yml differently? That is what it seems to be happening for me.

When using ${} in docker-compose.yml I get the following when running 'docker-compose build':
WARNING: compose.config.environment.__getitem__: The MYSQL_DB_PORT variable is not set. Defaulting to a blank string.
WARNING: compose.config.environment.__getitem__: The APP_NAME variable is not set. Defaulting to a blank string.
WARNING: compose.config.environment.__getitem__: The projectname variable is not set. Defaulting to a blank string.
WARNING: compose.config.environment.__getitem__: The APP_PORT variable is not set. Defaulting to a blank string.
ERROR: compose.cli.main.main: The Compose file './docker-compose.yml' is invalid because:
services.db.ports is invalid: Invalid port ":", should be [[remote_ip:]remote_port[-remote_port]:]port[/protocol]
services.rails.ports is invalid: Invalid port ":", should be [[remote_ip:]remote_port[-remote_port]:]port[/protocol]

When using $${} the build succeeds. But, when I try to run 'docker-compose run' I get the following:
Pulling db (mariadb:latest)...
^CGracefully stopping... (press Ctrl+C again to force)
Davids-iMac:projectname username$ dcompose --verbose up db
compose.config.config.find: Using configuration files: ./docker-compose.yml
docker.auth.find_config_file: Trying paths: ['/Users/username/.docker/config.json', '/Users/username/.dockercfg']
docker.auth.find_config_file: No config file found
compose.cli.command.get_client: docker-compose version 1.12.0, build b31ff33
docker-py version: 2.2.1
CPython version: 2.7.12
OpenSSL version: OpenSSL 1.0.2j  26 Sep 2016
compose.cli.command.get_client: Docker base_url: http+docker://localunixsocket
compose.cli.command.get_client: Docker version: KernelVersion=4.9.21-moby, Arch=amd64, BuildTime=2017-04-11T20:55:05.654443025+00:00, ApiVersion=1.29, Version=17.05.0-ce-rc1, MinAPIVersion=1.12, GitCommit=2878a85, Os=linux, Experimental=True, GoVersion=go1.7.5
compose.cli.verbose_proxy.proxy_callable: docker info <- ()
compose.cli.verbose_proxy.proxy_callable: docker info -> {u'Architecture': u'x86_64',
 u'BridgeNfIp6tables': True,
 u'BridgeNfIptables': True,
 u'CPUSet': True,
 u'CPUShares': True,
 u'CgroupDriver': u'cgroupfs',
 u'ClusterAdvertise': u'',
 u'ClusterStore': u'',
 u'ContainerdCommit': {u'Expected': u'9048e5e50717ea4497b757314bad98ea3763c145',
                       u'ID': u'9048e5e50717ea4497b757314bad98ea3763c145'},
...
compose.cli.verbose_proxy.proxy_callable: docker inspect_network <- (u'projectname_default')
compose.cli.verbose_proxy.proxy_callable: docker inspect_network -> {u'Attachable': True,
 u'Containers': {},
 u'Created': u'2017-04-16T02:34:45.872086431Z',
 u'Driver': u'bridge',
 u'EnableIPv6': False,
 u'IPAM': {u'Config': [{u'Gateway': u'172.23.0.1',
                        u'Subnet': u'172.23.0.0/16'}],
           u'Driver': u'default',
           u'Options': None},
 u'Id': u'bbce2ad83906ffcfdda42cbc2f8d7a53b170188a246657462feb7c5e198b9b95',
...
compose.cli.verbose_proxy.proxy_callable: docker containers <- (all=False, filters={u'label': [u'com.docker.compose.project=projectname', u'com.docker.compose.oneoff=False']})
compose.cli.verbose_proxy.proxy_callable: docker containers -> (list with 0 items)
compose.cli.verbose_proxy.proxy_callable: docker containers <- (all=True, filters={u'label': [u'com.docker.compose.project=projectname', u'com.docker.compose.service=db', u'com.docker.compose.oneoff=False']})
compose.cli.verbose_proxy.proxy_callable: docker containers -> (list with 0 items)
compose.cli.verbose_proxy.proxy_callable: docker inspect_image <- ('mariadb:latest')
compose.service.pull: Pulling db (mariadb:latest)...
compose.cli.verbose_proxy.proxy_callable: docker pull <- (u'mariadb', tag=u'latest', stream=True)
docker.auth.get_config_header: Looking for auth config
docker.auth.get_config_header: No auth config in memory - loading from filesystem
docker.auth.find_config_file: Trying paths: ['/Users/username/.docker/config.json', '/Users/username/.dockercfg']
docker.auth.find_config_file: No config file found
docker.auth.resolve_authconfig: Looking for auth entry for 'docker.io'
docker.auth.resolve_authconfig: No entry found
docker.auth.get_config_header: No auth config found
compose.cli.verbose_proxy.proxy_callable: docker pull -> <generator object _stream_helper at 0x10706a0a0>
latest: Pulling from library/mariadb
6d827a3ef358: Already exists
ed0929eb7dfe: Pull complete
03f348dc3b9d: Pull complete
fd337761ca76: Pull complete
f124cc324960: Pull complete
425010f95173: Pull complete
bd8823bd18e5: Pull complete
8122e5cee801: Pull complete
4cd5cdd269c6: Pull complete
d08d5473e577: Pull complete
8459510bffd2: Pull complete
bebdb1229b44: Pull complete
Digest: sha256:42bbb87f580850e60a3389a73a2e15f4c9ba19735304d1f157c213a3e579b8a2
Status: Downloaded newer image for mariadb:latest
compose.cli.verbose_proxy.proxy_callable: docker containers <- (all=True, filters={u'label': [u'com.docker.compose.project=projectname', u'com.docker.compose.service=db', u'com.docker.compose.oneoff=False']})
compose.cli.verbose_proxy.proxy_callable: docker containers -> (list with 0 items)
compose.parallel.feed_queue: Pending: set([<Service: db>])
compose.parallel.feed_queue: Starting producer thread for <Service: db>
compose.cli.verbose_proxy.proxy_callable: docker inspect_image <- ('mariadb:latest')
compose.cli.verbose_proxy.proxy_callable: docker inspect_image -> {u'Architecture': u'amd64',
 u'Author': u'',
 u'Comment': u'',
 u'Config': {u'ArgsEscaped': True,
             u'AttachStderr': False,
             u'AttachStdin': False,
             u'AttachStdout': False,
             u'Cmd': [u'mysqld'],
             u'Domainname': u'',
             u'Entrypoint': [u'docker-entrypoint.sh'],
...
compose.cli.verbose_proxy.proxy_callable: docker containers <- (all=True, filters={u'label': [u'com.docker.compose.project=projectname', u'com.docker.compose.service=db', u'com.docker.compose.oneoff=False']})
compose.cli.verbose_proxy.proxy_callable: docker containers -> (list with 0 items)
compose.cli.verbose_proxy.proxy_callable: docker inspect_image <- ('mariadb:latest')
compose.cli.verbose_proxy.proxy_callable: docker inspect_image -> {u'Architecture': u'amd64',
 u'Author': u'',
 u'Comment': u'',
 u'Config': {u'ArgsEscaped': True,
             u'AttachStderr': False,
             u'AttachStdin': False,
             u'AttachStdout': False,
             u'Cmd': [u'mysqld'],
             u'Domainname': u'',
             u'Entrypoint': [u'docker-entrypoint.sh'],
...
compose.service.build_container_labels: Added config hash: 9fd5aabce57b211eaf619f122608f28a51a845c0894e457ade3576c75cc29797
compose.cli.verbose_proxy.proxy_callable: docker create_host_config <- (memswap_limit=None, mem_swappiness=None, links=[], oom_score_adj=None, dns_search=None, pid_mode=None, log_config={'Type': u'', 'Config': {}}, cpu_quota=None, read_only=None, dns=None, volumes_from=[], ipc_mode=None, port_bindings={'${MYSQL_DB_PORT}/tcp': ['${MYSQL_DB_PORT}']}, security_opt=None, extra_hosts=None, dns_opt=None, cgroup_parent=None, network_mode=u'projectname_default', shm_size=None, userns_mode=None, tmpfs=None, mem_reservation=None, cap_drop=None, cap_add=None, restart_policy=None, devices=None, privileged=False, binds=[u'/Users/username/docker_images/projectname/.data/db:/var/lib/mysql:rw'], sysctls=None, pids_limit=None, mem_limit=None, group_add=None, ulimits=None)
compose.cli.verbose_proxy.proxy_callable: docker create_host_config -> {'Binds': [u'/Users/username/docker_images/projectname/.data/db:/var/lib/mysql:rw'],
 'Links': [],
 'LogConfig': {'Config': {}, 'Type': u''},
 'NetworkMode': u'projectname_default',
 'PortBindings': {'${MYSQL_DB_PORT}/tcp': [{'HostIp': '',
                                            'HostPort': '${MYSQL_DB_PORT}'}]},
 'VolumesFrom': []}
compose.service.create_container: Creating projectname_db_1
compose.cli.verbose_proxy.proxy_callable: docker create_container <- (name=u'projectname_db_1', image='mariadb:latest', labels={u'com.docker.compose.service': u'db', u'com.docker.compose.project': u'projectname', u'com.docker.compose.config-hash': '9fd5aabce57b211eaf619f122608f28a51a845c0894e457ade3576c75cc29797', u'com.docker.compose.version': u'1.12.0', u'com.docker.compose.oneoff': u'False', u'com.docker.compose.container-number': '1'}, host_config={'NetworkMode': u'projectname_default', 'Links': [], u'Isolation': None, 'PortBindings': {'${MYSQL_DB_PORT}/tcp': [{'HostPort': '${MYSQL_DB_PORT}', 'HostIp': ''}]}, 'Binds': [u'/Users/username/docker_images/projectname/.data/db:/var/lib/mysql:rw'], 'LogConfig': {'Type': u'', 'Config': {}}, 'VolumesFrom': []}, environment=[u'MYSQL_ROOT_PASSWORD=password', u'MYSQL_PASSWORD=password', u'MYSQL_USER=www', u'MYSQL_DB_HOST=db', u'MYSQL_DB_PORT=3306'], volumes={u'/var/lib/mysql': {}}, detach=True, ports=[(u'${MYSQL_DB_PORT}', u'tcp')], networking_config={u'EndpointsConfig': {u'projectname_default': {u'IPAMConfig': {}, u'Aliases': ['db']}}})
compose.parallel.parallel_execute_iter: Failed: <Service: db>
compose.parallel.feed_queue: Pending: set([])

ERROR: for db  Cannot create container for service db: invalid port specification: "${MYSQL_DB_PORT}"
ERROR: compose.cli.main.main: Encountered errors while bringing up the project.

**Comment by shin-** [comment]
Do you mind sharing your `docker-compose.yml` with us, as well as the output of the `env` command? Do you use a `.env` file?

**Comment by davidoshiro** [comment]
Shin,

I was not using a .env file. I believe I have figured out what was happening and it is pretty confusing. I was under the impression that declaring the environment variables in the docker-compose.yml file would allow me to reference them in the various config entries as well as in the Dockerfile.

I have since moved all environment variables to the .env file and have referenced them in Dockerfile via ARG. My intent was to consolidate each project's info in one place and use template Dockerfile and docker-compose.yml files. I believe my current setup is sufficient as I only need to update the .env file.
